File: leetcode/groupAnagrams.js
// https://leetcode.com/problems/group-anagrams/
// 
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.
// 
// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
//  typically using all the original letters exactly once.
//  
// 
// Example 1:
// Input: strs = ["eat","tea","tan","ate","nat","bat"]
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// Example 2:
// Input: strs = [""]
// Output: [[""]]
//
// Example 3:
// Input: strs = ["a"]
// Output: [["a"]]


/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
    const map = new Map();

    for (const word of strs) {

        const key = generateKey(word);

        if (!map.has(key)) {
            map.set(key, []);
        }

        map.get(key).push(word);
    }

    return Array.from(map.values());
};

function generateKey(word) {
    return word.split('').sort().join('');
}

// Runtime: 198 ms, faster than 45.15% of JavaScript online submissions for Group Anagrams.
// Memory Usage: 54.3 MB, less than 35.74% of JavaScript online submissions for Group Anagrams.


module.exports = {
    groupAnagrams
}

File: leetcode/algorithms/Algorithms/49-Group-Anagrams/GroupAnagrams.cs
/*
    Source: https://leetcode.com/problems/group-anagrams
    Runtime: 290 ms, faster than 33.72% of C# online submissions for Group Anagrams.
    Memory Usage: 50.3 MB, less than 65.23% of C# online submissions for Group Anagrams.
    Author: berkslv
    Date: 15/06/2022
*/
using System;
using System.Collections.Generic;

public class Solution049
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        var res = new List<IList<String>>();
        
        if (strs == null || strs.Length == 0) return res;
        
        var map = new Dictionary<String, List<String>>();
        
        for (int i = 0; i < strs.Length; i++)
        {
            char[] charArr = strs[i].ToCharArray();
            Array.Sort(charArr);
            var str = new String(charArr);
            if (map.ContainsKey(str))
            {
                map[str].Add(strs[i]);
            }
            else
            {
                var list = new List<String>();
                list.Add(strs[i]);
                map.Add (str, list);
            }
        }
        
        foreach (var iter in map.Values)
        {
            var item = (List<String>) iter;
            if (item.Count > 0) res.Add(item);
        }
        
        return res;
    }
}
